source/actions3/AMActionRunner3.cpp

```diff
@@ -476,6 +372,150 @@ void AMActionRunner3::onImmediateActionStateChanged(int state, int previousState
 	}
 }
 
+bool AMActionRunner3::isScanAction() const
+{
+	return qobject_cast<const AMScanAction *>(currentAction_) ? true : false;
+}
+
+void AMActionRunner3::setCurrentAction(AMAction3 *newAction)
+{
+	if(currentAction_ == newAction)
+		return;
+
+	// try to delete the old action
+	if (currentAction_) {
+		disconnect(currentAction_, 0, this, 0);
+		currentAction_->scheduleForDeletion();
+	}
+
+	// reset the current action
+	currentAction_ = newAction;
+	if (currentAction_) {
+		connect(currentAction_, SIGNAL(stateChanged(int,int)), this, SLOT(onCurrentActionStateChanged(int,int)));
+		connect(currentAction_, SIGNAL(progressChanged(double,double)), this, SIGNAL(currentActionProgressChanged(double,double)));
+		connect(currentAction_, SIGNAL(statusTextChanged(QString)), this, SIGNAL(currentActionStatusTextChanged(QString)));
+		connect(currentAction_, SIGNAL(expectedDurationChanged(double)), this, SIGNAL(currentActionExpectedDurationChanged(double)));
+	}
+
+	emit currentActionChanged(currentAction_);
+}
+
+bool AMActionRunner3::validateAction(AMAction3 *newAction, const QString &okButtonText, const QString &cancelButtonText )
+{
+	AMAction3::ActionValidity actionValidity = newAction->isValid();
+	if(actionValidity != AMAction3::ActionCurrentlyValid){
+		QString title;
+		QString message;
+		QString information;
+
+		if(actionValidity == AMAction3::ActionNotCurrentlyValid){
+			title = "The Action You've Added May Not Be Valid";
+			information = "\n\nYou can ignore this warning (maybe something will happen between now and the time the action is run to make it valid) or you can cancel adding the action.\nConsidering the action is about to run right now, this seems very unlikely.";
+		}
+		else if(actionValidity == AMAction3::ActionNeverValid){
+			title = "The Action You've Added Is Not Valid";
+			information = "\n\nYou cannot ignore this warning, this type of action is not supported.";
+		} else {
+			title = "The Action You've Added Is Valid";
+			information = "\n\nYou won't see this message.";
+		}
+
+		message = "The '" % newAction->info()->typeDescription() % "' action you're about to run reports:\n" % newAction->notValidWarning();
+
+		bool confirmToAddAction = AMErrorMon::showMessageBox(title, message, information, okButtonText, cancelButtonText, false); //default cancel
+		if (!confirmToAddAction) { // cancel butttton is clicked
+			cancelCurrentAction();
+			return false;
+		}
+	}
+
+	return true;
+}
+
+void AMActionRunner3::onCurrentActionStarting()
+{
+	if(loggingDatabase_)
```

Comment:
I'm sure that this was in the old code and I've just forgotten, but I just wanted to double check: is this essentially saying that logging is being initialized for this action?